### PR TITLE
AveragedField

### PR DESCRIPTION
**Pull request by glwagner**
This PR introduces an `AveragedField` type.

The intent is that this field will eventually replace `Average`. Before that occurs, we need to overhaul both output writers, but especially the NetCDFOutputWriter, so that they can properly slice all types of fields.

An `AveragedField` is constructed with the syntax (for example)

```julia
avg = AveragedField(operand; dims=(1, 2))
```

where `operand` is the object being averaged (which should also be a field), and `dims` are the dimensions over which the field is reduced. There is an optional argument `data` which would be an `OffsetArray` of the correct size which avoids extra memory allocation. We also overload `Statistics.mean`, so that the above code can be written

```julia
avg = mean(operand; dims=(1, 2))
```

When we have `ComputedField`s, `operand` can involve `AbstractOperations`. This will come in a future PR.

### CODE REVIEW

File: src/Fields/abstract_field.jl [hunk 1]
```diff
@@ -30,11 +31,34 @@ function validate_field_data(X, Y, Z, data, grid)
...
+
+Call compute! on fields after defining them.
+"""
+macro compute(def)
+    expr = Expr(:block)
+    field = def.args[1]
+    push!(expr.args, :($(esc(def))))
+    push!(expr.args, :(compute!($(esc(field)))))
+    return expr
+end
```
Comment (ali-ramadhan) [inline src/Fields/abstract_field.jl:57]: Kinda against defining this macro but we already discussed this on Slack. It seems appropriate for post-processing, which should become more possible with the fields abstraction so let's keep it.

File: src/Fields/averaged_field.jl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    avg.data.parent ./= prod(sz[d] for d in avg.dims)
+
+    return nothing
+end
+
+#####
+##### Very sugar
+#####
+
+Statistics.mean(ϕ::AbstractField; kwargs...) = AveragedField(ϕ; kwargs...)
```
Comment (ali-ramadhan) [inline src/Fields/averaged_field.jl:63]: 🏆

File: src/Fields/averaged_field.jl [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+#####
+##### Adapt
+#####
+
+Adapt.adapt_structure(to, averaged_field::AveragedField{X, Y, Z}) where {X, Y, Z} = 
+    AveragedField{X, Y, Z}(Adapt.adapt(to, averaged_field.data),
+                           Adapt.adapt(to, averaged_field.grid),
+                           Adapt.adapt(to, averaged_field.dims),
+                           Adapt.adapt(to, averaged_field.operand))
```
Comment (ali-ramadhan) [inline src/Fields/averaged_field.jl:73]: Do we know whether this `Adapt` rule works? It should, I see no reason why it wouldn't but curious if there's a test that uses it. I guess it has to be passed into a GPU kernel. Maybe that'll come in a future PR once the `AveragedField` sees more use.

File: test/test_diagnostics.jl [hunk 4]
```diff
@@ -13,23 +13,23 @@ function run_horizontal_average_tests(arch, FT)
 
     T̅ = Average(model.tracers.T, dims=(1, 2), with_halos=false)
     computed_T_profile = T̅(model)
-    @test size(computed_T_profile) == (1, 1, Nz)
-    @test computed_T_profile ≈ znodes(Cell, grid, reshape=true)
+    @test_skip size(computed_T_profile) == (1, 1, Nz)
+    @test_skip computed_T_profile ≈ znodes(Cell, grid, reshape=true)
```
Comment (ali-ramadhan) [inline test/test_diagnostics.jl:17]: Do these tests fail right now or are you skipping them in preparation for `Average` to be overhauled?

**Review by ali-ramadhan — APPROVED** [review]

File: src/Fields/abstract_field.jl (hunk 1 above)
Comment (glwagner) [reply src/Fields/abstract_field.jl:57]: We can open an issue to remove it after the PR is merged?

File: src/Fields/averaged_field.jl (hunk 3 above)
Comment (glwagner) [reply src/Fields/averaged_field.jl:73]: The tests for the subsequent PR #931 show that it does indeed work.

File: test/test_diagnostics.jl (hunk 4 above)
Comment (glwagner) [reply test/test_diagnostics.jl:17]: I am skipping them because the tests started timing out.

File: src/Fields/abstract_field.jl (hunk 1 above)
Comment (ali-ramadhan) [reply src/Fields/abstract_field.jl:57]: Nah we should keep it and see how it gets used. It's already useful in the tests. I might just have a personal prejudice against macros.

File: test/test_diagnostics.jl (hunk 4 above)
Comment (ali-ramadhan) [reply test/test_diagnostics.jl:17]: Hmmm, these were small fast tests and most of the setup still occurs but we can re-enable once testing doesn't time out.

Comment (glwagner) [reply test/test_diagnostics.jl:17]: Ah, I didn't look at the run time. I just noticed that adding the (largely identical) tests for `AveragedField` made the integration group time out. So I skipped these, and the tests stopped timing out. But maybe I just got lucky?

**Pull request merged**